cmd/vendorcycle/vendorcycle.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Checks for restricted imports in vendored dependencies.
+package main
+
+import (
+	"flag"
+	"fmt"
+	"go/build"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+)
+
+var (
+	exclude  = flag.String("exclude", `a^`, "skip packages pattern")
+	restrict = flag.String("restrict", `a^`, "restricted import pattern")
+)
+
+func main() {
+	flag.Parse()
+
+	args := flag.Args()
+
+	if len(args) == 0 {
+		log.Fatalf("usage: vendorcycle <dir>")
+	}
+	excludePattern, err := regexp.Compile(*exclude)
+	if err != nil {
+		log.Fatalf("Error compiling excluded package regex: %v", err)
+	}
+	importsPattern, err := regexp.Compile(*restrict)
+	if err != nil {
+		log.Fatalf("Error compiling excluded package regex: %v", err)
+	}
+
+	violations := map[string][]string{}
+	err = filepath.Walk(args[0], func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			p, err := build.Default.Import(fmt.Sprintf("./%s", path), ".", 0)
```

Comment:
fun thought: build tags?
I _think_ we're fine, the default tags should get setup in the script below, and non-default build tags shouldn't be a problem, probably?

Reply:
That is a valid point. I'm not sure how often we will hit it in practice, but I think regardless it might be better to refactor this to use https://godoc.org/golang.org/x/tools/go/packages

Reply:
Interesting observation, using `packages`, even with only `NeedImports` mode set is _significantly_ slower, which makes me think that it is actually pulling the modules rather than inspecting the source (this may have already been known by others :sweat_smile:). If we can afford to only inspect the source rather than loading the packages then that would certainly be advantageous as demonstrated by the 2 second `verify.no-vendor-cycles` run on the successful `verify-master` here.